circe-akka-serializer/src/main/scala/org/virtuslab/ash/CirceAkkaSerializer.scala

```diff
@@ -0,0 +1,82 @@
+package org.virtuslab.ash
+
+import akka.actor.ExtendedActorSystem
+import akka.event.{Logging, LoggingAdapter}
+import akka.serialization.SerializerWithStringManifest
+import io.circe.jawn.JawnParser
+import io.circe.{Decoder, Encoder, Printer}
+import org.reflections8.Reflections
+
+import java.io.NotSerializableException
+import java.nio.charset.StandardCharsets.UTF_8
+import java.util.NoSuchElementException
+import scala.jdk.CollectionConverters._
+import scala.reflect.runtime.{universe => ru}
+import scala.reflect.{ClassTag, classTag}
+
+abstract class CirceAkkaSerializer[Ser: ClassTag](actorSystem: ExtendedActorSystem)
+    extends SerializerWithStringManifest {
+  override val identifier = 42352
+  protected val logger: LoggingAdapter = Logging(actorSystem, this.getClass)
+
+  val codecs: Seq[(ru.TypeTag[_ <: Ser], (Encoder[_ <: Ser], Decoder[_ <: Ser]))]
+  val manifestMigrations: Seq[(String, String)]
+
+  private lazy val parents = codecs.flatMap { x =>
+    val clazz = x._1.tpe.typeSymbol.asClass
+    val mirror = ru.runtimeMirror(getClass.getClassLoader)
+    val rootClazzName = mirror.runtimeClass(clazz).getName
+    def getAllSubclasses(clazz: ru.ClassSymbol): List[ru.ClassSymbol] = {
+      if (!clazz.isSealed)
+        List(clazz)
+      else
+        clazz :: clazz.knownDirectSubclasses.toList.flatMap(x => getAllSubclasses(x.asClass))
+    }
+    getAllSubclasses(clazz).map(x => (mirror.runtimeClass(x).getName, rootClazzName))
+  }.toMap
+  private lazy val codecsMap = codecs.map(x => x.copy(_1 = x._1.tpe.typeSymbol.fullName)).toMap
+  private lazy val manifestMap = manifestMigrations.toMap
+
+  private val parser = new JawnParser
+  private val printer = Printer.noSpaces
+
+  override def manifest(o: AnyRef): String = parents.getOrElse(o.getClass.getName, "")
+
+  override def toBinary(o: AnyRef): Array[Byte] = {
+    codecsMap.get(manifest(o)) match {
+      case Some((encoder, _)) =>
+        val buffer = printer.printToByteBuffer(encoder.asInstanceOf[Encoder[AnyRef]](o), UTF_8)
+        val res = new Array[Byte](buffer.remaining)
+        buffer.get(res)
+        res
+      case None =>
+        throw new RuntimeException(
+          s"Serialization of ${o.getClass.getName} failed. Call Register[A] for this class or its supertype and append result to codecs.")
+    }
+  }
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = {
+    codecsMap.get(manifestMap.getOrElse(manifest, manifest)) match {
+      case Some((_, decoder)) =>
+        parser.parseByteArray(bytes).flatMap(_.as(decoder)).fold(e => throw e, identity).asInstanceOf[AnyRef]
+      case None =>
+        throw new NotSerializableException(
+          s"Manifest $manifest did not match any known codec. If you're not currently performing a rolling upgrade, you must add a manifest migration to correct codec.")
+    }
+  }
+
+  //noinspection SameParameterValue
+  protected def scanForSerializables(packagePrefix: String): Unit = {
```

Comment:
Oh so that's effectively a runtime implementation for https://github.com/VirtusLab/akka-serialization-helper/issues/57 right?

Reply:
Maybe this should be non-optional? As in, a call to `scanForSerializables` should be included within `CirceAkkaSerializer` c'tor, with `packagePrefix` as an abstract val to provided by subclasses? 🤔

Reply:
In other words, what could be a situation where someone would NOT like such check to be run? I can think of performance reasons (although that's one-off & negligible), or some unusual package structures

Reply:
I'm not sure what you mean, this function is called explicitly by the library user.
Example [here](https://github.com/VirtusLab/akka-serialization-helper/blob/311fa1ebd8bc04bf990e11f2940ecff9b234271d/circe-akka-serializer/src/test/scala/org/virtuslab/ash/CustomSerializer.scala#L22)